### PR TITLE
Merge csv outputs

### PR DESCRIPTION
**Pull request by ellisrichardj**
I've added an additional process to merge the csv output files and determine whether samples are M.bovis positive or not.  Still working on the unit test for this but thought that could be done in a separate branch.

### CODE REVIEW

File: bTB-WGS_process.nf [hunk 1]
```diff
@@ -320,9 +320,26 @@ process IDnonbovis{
...
+	
+	input:
+	file('Assigned.csv') from Assigned
+	file('Qbovis.csv') from Qbovis
+
+	output:
+	file '*.csv' into FinalOut
+
+	"""
+	python3 $pypath/CombineCsv.py Assigned.csv Qbovis.csv ${params.DataDir}
```
Comment (afishman) [inline bTB-WGS_process.nf:340]: This should be in `bin/` like the other processes. You can remove the `python3 $pypath/` part

File: pyscripts/CombineCsv.py [hunk 2]
```diff
@@ -0,0 +1,35 @@
+import pandas as pd
```
Comment (afishman) [inline pyscripts/CombineCsv.py:1]: Add `#!/usr/local/bin/python` to the top. As well as a brief description of what it does. Eg see https://github.com/APHA-CSU/BovTB-nf/blob/master/tests/utils/print_csv_value.py

File: pyscripts/CombineCsv.py [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+
+#Merge dataframes and fill ID with M bovis if Pass, then any remaining blank cells with 'NA'
+FinalOut = pd.merge(Assigned, Qbovis, on = 'Sample', how = 'outer')
+FinalOut['ID'].fillna('Mycobacterium bovis', inplace = True)
+FinalOut.fillna('NA', inplace = True)
+FinalOut.set_index('Sample', inplace = True)
+FinalOut.sort_index(inplace = True)
+
+#Write to csv
+FinalOut.to_csv(""+str(SeqRun)+"_FinalOut_"+str(DateOut)+".csv")
```
Comment (afishman) [inline pyscripts/CombineCsv.py:35]: It's convention to add ` if __name__ == '__main__':` here to make this module easier to import. See: https://stackoverflow.com/questions/419163/what-does-if-name-main-do

File: pyscripts/CombineCsv.py [hunk 4]
```diff
@@ -0,0 +1,35 @@
+import pandas as pd
+import sys, os
+from datetime import datetime, date
+
+In1 = sys.argv[1]
```
Comment (afishman) [inline pyscripts/CombineCsv.py:5]: I'd put arg parsing in the `if __name__ == '__main__'` section. Sometimes you might want to use inputs that don't come from the command-line. 

You could also try using [argparse ](https://github.com/APHA-CSU/BovTB-nf/blob/master/tests/utils/print_csv_value.py)if you wanted - but that's less of a big deal.

File: pyscripts/CombineCsv.py [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+In2 = sys.argv[2]
+SeqRun = sys.argv[3]
+
```
Comment (afishman) [inline pyscripts/CombineCsv.py:8]: The rest of the script you can probably wrap up into a function. E.g. `def combine(assigned, csv_path, sequence_run)`

File: pyscripts/CombineCsv.py [hunk 6]
```diff
@@ -0,0 +1,35 @@
...
+DateOut = date.today().strftime('%d%b%y')
```
Comment (afishman) [inline pyscripts/CombineCsv.py:9]: We should be consistent with our naming conventions. The python code I wrote in `tests/utils/` follows the pep8 standard: the common standard. https://www.python.org/dev/peps/pep-0008/

Variable names should be lowercase underscore. E.g `date_out`

File: pyscripts/CombineCsv.py [hunk 7]
```diff
@@ -0,0 +1,35 @@
...
+
+# Read BovPos csv and change ID to either 'Negative' or 'Mycobacterium bovis',
+# depending on number and relative abundance of M.bovis reads.
+# Thresholds set by comparison with output from spoligotyping.
+# Fill empty value cells with zero 
+Qbovis = pd.read_csv(In2)
+Qbovis.loc[(Qbovis['TotalReads'] >= 500) & (Qbovis['Abundance'] >= 1), 'ID'] = 'Mycobacterium bovis'
+Qbovis.loc[(Qbovis['TotalReads'] < 500) & (Qbovis['Abundance'] < 1), 'ID'] = 'Negative'
+Qbovis.loc[(Qbovis['TotalReads'] < 500) & (Qbovis['Abundance'] > 1), 'ID'] = 'Inconclusive'
+Qbovis.loc[(Qbovis['TotalReads'] > 500) & (Qbovis['Abundance'] < 1), 'ID'] = 'Inconclusive'
```
Comment (afishman) [inline pyscripts/CombineCsv.py:22]: You should DRY up this hardcoded value. A couple of option for this:
- a global variable just below the imports called something like `threshold = 500`.
- An optional argument. `def combine(assigned, csv_path, sequence_run, threshold=500)`
- An optional command-line argument. `parser.add_argument('threshold', default=500, help='blah')`

File: pyscripts/CombineCsv.py (hunk 3 above)
Comment (afishman) [inline pyscripts/CombineCsv.py:35]: Can we perhaps use a simpler csv naming convention?

**Review by afishman — COMMENTED** [review]
overall approach seems good. Most of the feedback is on style an conventions. It'd be great to see some unit tests as well. This is a nice warmup problem

File: bTB-WGS_process.nf (hunk 1 above)
Comment (ellisrichardj) [reply bTB-WGS_process.nf:340]: Done

File: pyscripts/CombineCsv.py (hunk 3 above)
Comment (ellisrichardj) [reply pyscripts/CombineCsv.py:35]: I need the csv name to reflect show the seqneucer run number and the date the output was processed.  This avoids conflicts if the same data is processed on different days

**Comment by ellisrichardj** [comment]
Refactored CombineCsv.py to address your comments - thanks, I learnt a lot!

File: bTB-WGS_process.nf [hunk 8]
```diff
@@ -320,9 +320,26 @@ process IDnonbovis{
...
+	
+	input:
+	file('Assigned.csv') from Assigned
+	file('Qbovis.csv') from Qbovis
+
+	output:
+	file '*.csv' into FinalOut
+
+	"""
+	CombineCsv.py Assigned.csv Qbovis.csv ${params.DataDir}
```
Comment (afishman) [inline bTB-WGS_process.nf:340]: Can you rename this to combineCsv.py? I know it sounds picky but that's the convention in `bin/`

File: bin/CombineCsv.py [hunk 9]
```diff
@@ -0,0 +1,49 @@
+#!/usr/bin/env python3
+
+# Merge the csv output files and determine whether samples are M.bovis positive or not, based on predetemined 
+# thresholds for number and proportion of M.bovis reads.
+
+import pandas as pd
+import argparse
+from datetime import datetime, date
+
+date_out = date.today().strftime('%d%b%y')
```
Comment (afishman) [inline bin/CombineCsv.py:10]: This should go in `combine()`.

Currently this line runs once on module import. If someone imports this module runs `combine()` multiple times, `date_out` won't update

File: bin/CombineCsv.py [hunk 10]
```diff
@@ -0,0 +1,49 @@
...
+
+def combine(assigned_csv, bovis_csv, seq_run, read_threshold, abundance_threshold):
+    #Read Assigned Clade csv and replace blank cells  with 'NA'
+    assigned_df = pd.read_csv(assigned_csv)
+
+    # Read BovPos csv and change ID to either 'Negative' or 'Mycobacterium bovis',
+    # depending on number and relative abundance of M.bovis reads.
+    # Thresholds set by comparison with output from spoligotyping.
+    # Fill empty value cells with zero
```
Comment (afishman) [inline bin/CombineCsv.py:19]: Can you add a brief description of the bug we talked over the phone? (as well as any others you think may also exist here). 
Convention I'm used to is if there's a bug, or shortcoming you know about but don't want to fix yet, you should add a searchable comment like 
`BUG: Might give spurious results if <blah>`
or 
`TODO: default 'ID' required to cover all logical cases`

File: bin/CombineCsv.py [hunk 11]
```diff
@@ -0,0 +1,49 @@
...
+
+    #Write to csv
+    finalout_df.to_csv(""+str(seq_run)+"_FinalOut_"+str(date_out)+".csv")
+
+if __name__ == '__main__':
+    parser = argparse.ArgumentParser()
+    parser.add_argument('assigned_csv', help='path to AssignedWGSClade.csv')
+    parser.add_argument('bovis_csv', help='path to Bovis.csv')
+    parser.add_argument('seq_run', help='Unique sequencer run number')
+    parser.add_argument('read_threshold', nargs='?', type=int, default=500, help='threshold for number of M.bovis reads')
```
Comment (afishman) [inline bin/CombineCsv.py:44]: FYI another way to do optional arguments on argparse is with something like 
`parser.add_argument('--read_threshold', type=int, default=500, help='threshold for number of M.bovis reads')`

Though I'm fine with what you prefer

File: bin/CombineCsv.py [hunk 12]
```diff
@@ -0,0 +1,49 @@
...
+
+    #Merge dataframes and fill ID with M bovis if Pass, then any remaining blank cells with 'NA'
+    finalout_df = pd.merge(assigned_df, qbovis_df, on = 'Sample', how = 'outer')
+    finalout_df['ID'].fillna('Mycobacterium bovis', inplace = True)
+    finalout_df.fillna('NA', inplace = True)
+    finalout_df.set_index('Sample', inplace = True)
+    finalout_df.sort_index(inplace = True)
+
+    #Write to csv
+    finalout_df.to_csv(""+str(seq_run)+"_FinalOut_"+str(date_out)+".csv")
```
Comment (afishman) [inline bin/CombineCsv.py:37]: Two more readable ways to write this:
- `"%s_FinalOut_%s.csv" % (seq_run, date_out)`
- `"{}_FinalOut_{}.csv".format(seq_run, date_out)`

**Review by afishman — CHANGES_REQUESTED** [review]

File: bin/CombineCsv.py (hunk 9 above)
Comment (ellisrichardj) [reply bin/CombineCsv.py:10]: Done

File: bin/CombineCsv.py (hunk 12 above)
Comment (ellisrichardj) [reply bin/CombineCsv.py:37]: Done - second option

File: bin/CombineCsv.py (hunk 11 above)
Comment (ellisrichardj) [reply bin/CombineCsv.py:44]: Thanks - makes sense

Comment (ellisrichardj) [reply bin/CombineCsv.py:44]: Thanks, I like this version

File: bin/CombineCsv.py (hunk 10 above)
Comment (ellisrichardj) [reply bin/CombineCsv.py:19]: Added in the comment to describe the issue

File: bTB-WGS_process.nf (hunk 8 above)
Comment (ellisrichardj) [reply bTB-WGS_process.nf:340]: Done

**Comment by ellisrichardj** [comment]
Updated as requested

**Review by afishman — APPROVED** [review]
Excellent bit of work. Looks very nice. Well done. 🙂

**Pull request merged**